Global benchmark

SichongP opened the pull request:
### Description

This PR adds two features:  
- `--disable-benchmark` to turn off benchmarking for all rules  
- `--workflow-benchmark FILE` to force benchmarking all rules and output all stats to a single file  

Related issues:  
https://github.com/snakemake/snakemake/issues/764  
https://github.com/snakemake/snakemake/issues/564
https://github.com/snakemake/snakemake/issues/50
### QC
<!-- Make sure that you can tick the boxes below. -->

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

SichongP commented:
I noticed a strange behavior when working on this PR:

If the target rule has `benchmark` directive and if the benchmark file exists, rerunning the workflow when target rule's input files are absent shows nothing to be done:

Snakefile:
```
rule all:
    input: "test.txt"
    benchmark: "all.benchmark"

rule A:
    output: "test.txt"
    shell: 
     """
     touch {output}
     """
```

```
> snakemake -j 1
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job      count    min threads    max threads
-----  -------  -------------  -------------
A            1              1              1
all          1              1              1
total        2              1              1

Select jobs to execute...

[Thu Mar 16 21:42:09 2023]
rule A:
    output: test.txt
    jobid: 1
    reason: Missing output files: test.txt
    resources: tmpdir=/tmp

[Thu Mar 16 21:42:09 2023]
Finished job 1.
1 of 2 steps (50%) done
Select jobs to execute...

[Thu Mar 16 21:42:09 2023]
localrule all:
    input: test.txt
    jobid: 0
    benchmark: all.benchmark
    reason: Missing output files: all.benchmark; Input files updated by another job: test.txt
    resources: tmpdir=/tmp

[Thu Mar 16 21:42:09 2023]
Finished job 0.
2 of 2 steps (100%) done
Complete log: .snakemake/log/2023-03-16T214208.618280.snakemake.log


> rm test.txt
> snakemake -j 1
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).
Complete log: .snakemake/log/2023-03-16T214310.806788.snakemake.log


> rm all.benchmark
> snakemake -j 1
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job      count    min threads    max threads
-----  -------  -------------  -------------
A            1              1              1
all          1              1              1
total        2              1              1

Select jobs to execute...

[Thu Mar 16 21:44:06 2023]
rule A:
    output: test.txt
    jobid: 1
    reason: Missing output files: test.txt
    resources: tmpdir=/tmp

[Thu Mar 16 21:44:06 2023]
Finished job 1.
1 of 2 steps (50%) done
Select jobs to execute...

[Thu Mar 16 21:44:06 2023]
localrule all:
    input: test.txt
    jobid: 0
    benchmark: all.benchmark
    reason: Missing output files: all.benchmark; Input files updated by another job: test.txt
    resources: tmpdir=/tmp

[Thu Mar 16 21:44:06 2023]
Finished job 0.
2 of 2 steps (100%) done
Complete log: .snakemake/log/2023-03-16T214406.375213.snakemake.log
```

Note: this error does not occur if target file is specified:
`snakemake -j 1 test.txt` or `snakemake -j 1 --until A`


As a result, when `--workflow-benchmark` is set, we will see a similar error upon rerun. I'm still trying to chase down the exact cause of this behaviour.

SichongP commented:
The test is failing. Likely related: https://github.com/conda-forge/libxml2-feedstock/pull/82

SichongP commented:
> If the target rule has benchmark directive and if the benchmark file exists, rerunning the workflow when target rule's input files are absent shows nothing to be done:

As I mentioned in https://github.com/snakemake/snakemake/issues/2142, the cause is https://github.com/snakemake/snakemake/blob/main/snakemake/dag.py#L1107. When either `log` or `benchmark` is specified, `job.has_products()` returns False.

fgvieira commented:
I'd just like to suggest maybe renaming `--disable-benchmark` to `--nobenchmark`, just to keep the same logic as (e.g.) `--notemp`.

SichongP commented:
> Very nice changes, speciall in adding `jobid`, `rule`, and `wildcards` to the final output file. If not too difficult, it would be great if you could also add to the global benchmark file:
> 
> * `threads` requested
> * resources requested
> * size of input file (e.g., `input.size_mb`)
> * [Benchmark to log real time of job #502](https://github.com/snakemake/snakemake/issues/502) asks for real job time; not sure why he can't use the first two columns but, in case it is something else, maybe also include it?
> 
> I'd just like to suggest maybe renaming `--disable-benchmark` to `--nobenchmark`, just to keep the same logic as (e.g.) `--notemp`.

Thanks for the suggestions! I've added threads, resources, and input size columns to final output. It's definitely helpful to have them in one place when benchmarking and optimizing a workflow.

Also changed `--disable-benchmark` to `--nobenchmark` for consistency.

fgvieira commented:
> Thanks for the suggestions! I've added threads, resources, and input size columns to final output. It's definitely helpful to have them in one place when benchmarking and optimizing a workflow.

That has been my problem when trying to optimize a workflow! Thanks a lot for your changes!!

As for the walltime issue, not sure what he is referring to, but I asked him to clarify.

VPestel commented:
Great feature @SichongP ! I was actually looking for that. I'm wondering if it couldn't be an occasion to fix a bug I reported few weeks ago about the CPU time measurement, that is flawed when executing multiple instruction in a `shell` block: #2147

SichongP commented:
> Great feature @SichongP ! I was actually looking for that. I'm wondering if it couldn't be an occasion to fix a bug I reported few weeks ago about the CPU time measurement, that is flawed when executing multiple instruction in a `shell` block: #2147

I think it's best to limit scope of a PR so it's easier to review and maintain. It seems that you already have a solution in mind though. Feel free to open a separate one for that particular fix. Thanks!

fgvieira commented:
hi @SichongP , thanks again for this PR!

As I see it, `size_mb` will be mostly used to try to predict memory usage. However, some programs do not load in memory all input files, but rather just some (e.g. a database). As such, I think it could also be useful to include, apart from `size_mb`, the size of each input file (e.g. comma-separated).

Not sure how feasible it is, but according to `snakemake` code it should not be too difficult:
https://github.com/snakemake/snakemake/blob/5e5cb619bda084baf159d1d52767459b4a2f257d/snakemake/io.py#L1655-L1662

Maybe something like:
```
_benchmark.insert(0, "input_size_files", ",".join([f.size / 1024 / 1024 for f in job.input]))
```

johanneskoester commented:
> @fgvieira Thanks for the suggestion! I've changed `input_size_mb` to report file size individually, with file name prepended for improved readability.
> 
> This does clutter the output by quite a bit for rules with many input files. A JSON or YAML format might be a better option. But since most snakemake users will already be comfortable parsing a tabular file format, I'm inclined to keep it as a tsv for now.

Mhm, changing it later will lead to a breaking change though. I think I would prefer JSON in this case. Or actually rather msgpack, in order to make it small, as it will contain lots of numbers which are binary encoded in msgpack.

fgvieira commented:
Sounds good by me.

SichongP commented:
> Mhm, changing it later will lead to a breaking change though. I think I would prefer JSON in this case. Or actually rather msgpack, in order to make it small, as it will contain lots of numbers which are binary encoded in msgpack.

I'm not familiar with msgpack but I wonder if the saving in file size is worth it if it's not as commonly used and therefore less use friendly? JSON seems like a good comprise -- plenty of people already familiar with it and it's easy to take a quick look at the text file without having to learn a new library.

johanneskoester commented:
That is a valid concern. I was actually thinking that msgpack is supported by pandas but this has been deprecated, already indicating that msgpack is not a good choice.

johanneskoester commented:
What about removing the input_size_mb column from the primary output table and instead record that in the regular persistence records for input files. One could then generate the complete table you want to have here on the fly via a command, instead of having to store this information in the TSV. In particular the input file sizes would be repeated in many rows in case a particular file is used by many jobs.

SichongP commented:
> What about removing the input_size_mb column from the primary output table and instead record that in the regular persistence records for input files. One could then generate the complete table you want to have here on the fly via a command, instead of having to store this information in the TSV. In particular the input file sizes would be repeated in many rows in case a particular file is used by many jobs.

So, `--workflow-benchmark FILE` would generate a tsv `FILE`, which doesn't have individual file size. Then a separate command (`snakemake --benchmark-file-size` maybe?) run on a previously completed workflow would give a list of all input files and their size, pulled from persistence records?

If they are decoupled, I worry that there is a good chance someone will be comparing input file sizes to an out-of-date benchmark file. And we have no way of guarding against that, no?

fgvieira commented:
What about if we encode them as JSON?

johanneskoester commented:
What about removing the input_size_mb column from the primary output table and instead record that in the regular persistence records for input files. One could then generate the complete table you want to have here on the fly via a command, instead of having to store this information in the TSV. In particular the input file sizes would be repeated in many rows in case a particular file is used by many jobs.

> > What about removing the input_size_mb column from the primary output table and instead record that in the regular persistence records for input files. One could then generate the complete table you want to have here on the fly via a command, instead of having to store this information in the TSV. In particular the input file sizes would be repeated in many rows in case a particular file is used by many jobs.
> 
> So, `--workflow-benchmark FILE` would generate a tsv `FILE`, which doesn't have individual file size. Then a separate command (`snakemake --benchmark-file-size` maybe?) run on a previously completed workflow would give a list of all input files and their size, pulled from persistence records?
> 
> If they are decoupled, I worry that there is a good chance someone will be comparing input file sizes to an out-of-date benchmark file. And we have no way of guarding against that, no?

I'd rather do the following:
```
snakemake --benchmark-all
```
records benchmark information for all jobs inside of the .snakemake folder, without the file size info.
And
```
snakemake --print-benchmark-all
```
Prints the full table including the file sizes to STDOUT, taking the info from what is stored in the .snakemake folder.
This way, we have the functionality you intend, but we also just have the bare necessities stored in .snakemake for using it in other ways (e.g. predicting mem_mb and runtime for subsequent runs).

SichongP commented:
PR has been updated based on discussion above. Key changes:
There are now 4 new command line options:
- `--benchmark-all FILE` will force benchmarking all rules (except default target rule) and write metrics (without individual input file sizes) to FILE in TSV format
- `--nobenchmark` will disable benchmarking for all rules
- `--print-benchmark-all` will print benchmark metrics to STDOUT from **all** rules. This requires a previous successful run **and** that all rules are benchedmarked with either `benchmark` directive or `--benchmark-all`
- `--print-benchmark` will print benchmark metrics to STDOUT from only rules with `benchmark` directive. This requires a previous successful run.

Both `--print-benchmark` and `--print-benchmark-all` will an additional column `input_sizes_mb` in addition to the same table output as `--benchmark-all`. `input_sizes_mb` column will contain sizes of input files for each rule, separated by `;`. File sizes are pulled from persistence record in `.snakemake/metadata/`

A new `input_sizes_mb` record is added to persistence metadata.

I added both `--print-benchmark` and `--print-benchmark-all` because I want to:
1. Support an easy way to aggregate benchmarks from desired rules without having to call `--benchmark-all` first. 
2. Institute a check to ensure that all benchmark files have been created, to avoid accidentally missing some jobs because the user does not remember that a previous run had failed

If one wants to specifically benchmark a section of workflow, they can do so by combining  `--print-benchmark` or `--print-benchmark-all` with target rule, target file, or `--until`

SichongP commented:
That failed test is passing in my local env. I ran into a similar problem yesterday with that `protobuf` warning. I think downgrading that package from 4.23.1 to 4.21.12 fixed it for me

fgvieira commented:
I find adding two new options (`--print-benchmark-all` and `--print-benchmark`) to just add a column to a TSV file, a bit overkill and confusing (why do they have to be run again after a successful run?).

What if we replace them by an option (e.g.) `--benchmark-format` where it can be specified how verbose the benchmark file is? We could have a number or predefined levels (e.g. `short`, `extended`, `full`).

Alternatively, we could also actually have different formats (this way, it can be easily extended in the future to other formats):
- `tsv`: same as this PR but without `input_sizes_mb`
- `json`: saves benchmarks in json format and includes everything (including `input_sizes_mb`)

SichongP commented:
> I find adding two new options (`--print-benchmark-all` and `--print-benchmark`) to just add a column to a TSV file, a bit overkill and confusing (why do they have to be run again after a successful run?).
> 
> What if we replace them by an option (e.g.) `--benchmark-format` where it can be specified how verbose the benchmark file is? We could have a number or predefined levels (e.g. `short`, `extended`, `full`).
> 
> Alternatively, we could also actually have different formats (this way, it can be easily extended in the future to other formats):
> 
> * `tsv`: same as this PR but without `input_sizes_mb`
> * `json`: saves benchmarks in json format and includes everything (including `input_sizes_mb`)

I agree that `--print-benchmark-all` and `--benchmark-all` feel a little redundant. `--print-benhcmark`, however, serves a distinct purpose -- to aggregate benchmark metrics from rules with `benchmark` directive in their rule definition. Sometimes I have a set of rules that I want to benchmark but I don't want to benchmark every single rule. Currently I have to use an additonal rule or some post-run script to gather all these metrics. Now `--print-benchmark` provides an easy way to aggregate them. We can discuss if `--print-benchmark` should be used concurrently with a run or it should be invoked post-run, but its functionality is useful, imo.

For `--benchmark-format [short, extended, full]` option, I don't see many benifits of it because we are not really offering THAT many metrics that anyone would want or need to customize what gets printed out (like `top` or slurm's `sinfo` do).

As to supporting different file format, I'm also not sure how useful that is -- TSV is generally a better choice for flat, tabular data. Our tsv output has consistent structure and therefore should be easy to parse, but also simple to just view from a text editor. JSON, on the other hand, is more suitable for more complex and nested data, and is not exactly easy to view as plain text. I'm not sure if there is a use case where JSON is needed and TSV can't achieve it. 

Perhaps we go back to the original approach, drop `--print-benchmark-all` and have `--benchmark-all` print out `input_sizes_mb` but as the last column? That way it can easily be cut or dropped if one doesn't want it, but still there for those that do?

fgvieira commented:
For me, I'd have all fields in a TSV generated concurrently with a run (i.e. not invoked post-run), since it is much simpler for the user (benchmark files always generated when the rule is successful, and contain all the info). As for the options, the only relevant option might be `--nobenchmark` (as far as I know benchmarking has a small overhead, so why not do it always?).

As I see it, `--print-benchmark-all` and `--print-benchmark` just aggregate results, which (if we have a TSV) can be done with `find`/`cat`. If you are interested on just a subset, grep on rule names.

As for `--benchmark-all FILE`, it benchmarks all rules, even those without the `benchmark` directive. It seems you want to have two sets of benchmarks, one for the rules with the directive and another with everything. But why not add the directive to all the rules you might want to benchmark and grep for the rules you need?

The JSON suggestion was just because @johanneskoester mentioned he would prefer it. But for me, TSV would be fine!

SichongP commented:
Ok how about:
1. `--benchmark-all FILE` benchmarks **all** rules and writes **all** metrics to FILE, including individual input file sizes
2. `--agg-benchmark FILE` (or `--print-benchmark FILE`, I'm okay with either name) writes benchmark of **benchmarked** rules (those with `benchmark` directive) and writes **all** metrics to FILE, including individual input file sizes
3. `--nobenchmark` disables `benchmark` directive (also disables `--benchmark-all` is someone for some reason uses both)

Both `--benchmark-all FILE` and `--agg-benchmark FILE` shuold be invoked with a run.

fgvieira commented:
For me, the most relevant ones would be:
1. `--benchmark FILE` (invoked with a run): writes benchmark of **benchmarked** rules and **all** metrics to FILE, including individual input file sizes
2. `--nobenchmark` disables all `benchmark` directive

But like I said before, nothing against extra options if someone finds them useful! So an extra one to all rules would also be fine:

3. `--benchmark-all FILE` (invoked with a run): benchmarks **all** rules and writes **all** metrics to FILE, including individual input file sizes

SichongP commented:
> For me, the most relevant ones would be:
> 
> 1. `--benchmark FILE` (invoked with a run): writes benchmark of **benchmarked** rules and **all** metrics to FILE, including individual input file sizes
> 2. `--nobenchmark` disables all `benchmark` directive
> 
> But like I said before, nothing against extra options if someone finds them useful! So an extra one to all rules would also be fine:
> 
> 3. `--benchmark-all FILE` (invoked with a run): benchmarks **all** rules and writes **all** metrics to FILE, including individual input file sizes

I like this setup! I have updated the PR to include these three arguments only

SichongP commented:
There seems to be some major refactoring happening in the main branch. @johanneskoester Is main okay to be merged back to my branch at the moment?

fgvieira commented:
@johanneskoester is the curent `master` branch relatively stable? If so, I think it would be nice to try to merge these changes as soon as possible.
@SichongP can you try to fix the conflicts?

SichongP commented:
> @johanneskoester is the curent `master` branch relatively stable? If so, I think it would be nice to try to merge these changes as soon as possible. @SichongP can you try to fix the conflicts?

I see a pending PR for v8.0.0 release. Maybe once that is merged I can try to merge main back into my branch? 

I haven't been following the refactoring super closely so I'd definitely appreciate any help with understanding the new structure!

corneliusroemer commented:
It would be great to have a simple global way to get a benchmark tsv without having to directive bloat, I'd be happy to test. v8 was released so nothing should be in the way of merging main back into this PR.

Regarding TSV vs JSON, let's remember that the issue/feature requests are all about making it easier to globally enable (or disable) benchmarks, not that the current benchmarks don't contain enough information or should be in a different format.

I suspect 90% of users who end up using this option will prefer TSV over JSON. It should be easy to extend to JSON _if_ it turns out that a lot of people have trouble parsing and/or want extra nested features.

corneliusroemer commented on snakemake/__init__.py:
```diff
@@ -633,18 +638,27 @@ def snakemake(
+        if benchmark_all is not None:
+            if not forceall:
+                logger.warning(
+                    "Warning: Benchmarking all rules but --forceall is not set. Resulting benchmarks maybe incomplete or outdated!"
```
I think this warning is unnecessary

SichongP replied:
How so? If a rule is skipped due to a previous run, but if hardware (eg. cloud platforms) have changed since then, current run would still grab metrics from previous files and thus they would be inaccurate, no?

corneliusroemer commented on snakemake/__init__.py:
```diff
@@ -334,6 +336,9 @@ def snakemake(
         keep_incomplete (bool):     keep incomplete output files of failed jobs
+        nobenchmark (bool):         Disable benchmarking for all rules
+        benchmark_output (str):            Write all benchmared stats to given file (tab-separated)
+        benchmark_all (str):        Benchmark all rules regardless if benchmark directive is set in Snakefile. Write all benchmark stats to given file (tab-separated)
```
It might be neat to automatically generate the benchmark filename if none is provided, somewhere in `.snakemake` with a datestamp of when the run was started.

That way one wouldn't need to change the name when rerunning and not wanting to overwrite an existing benchmark output.

SichongP replied:
Added

corneliusroemer commented on snakemake/benchmark.py:
```diff
@@ -386,3 +388,63 @@ def write_benchmark_records(records, path):
+                "Error: \n"
+                + f"Benchmark file {job._benchmark} does not "
+                + f"exist for rule {job.rule.name} \n"
+                + "A complete run is required to print all benchmark metrics."
```
Why would we require all jobs to complete? Can we not just leave the output empty for rules that failed, or include an extra column that states "failed/succeeded"? It would be good to know benchmark results for failed jobs as well, as they might have failed due to OOM or exceeding cluster QOS time. This would also serve as a quick way to grep all failed jobs.

SichongP replied:
I agree. But currently Snakemake does not write a benchmark file if a job fails and therefore I think it's a little out of scope for this PR to change that behavior.

corneliusroemer commented on snakemake/benchmark.py:
```diff
@@ -386,3 +388,63 @@ def write_benchmark_records(records, path):
+        wildcard_str = ";".join(
+            ["NA"]
+            if len(job.wildcards_dict) == 0
+            else [f"{name}={value}" for name, value in job.wildcards_dict.items()]
```
This simple encoding will not be parseable if any of name/value contain either of `;` or `=`

One could quote encode in those cases and escape quotes with backslash.

corneliusroemer commented on snakemake/benchmark.py:
```diff
@@ -386,3 +388,63 @@ def write_benchmark_records(records, path):
+            infile_sizes = persistence.input_sizes_mb(job.output)
+            input_file_size = (
+                ";".join(
+                    "{name}={size:0.2f}".format(
```
Might need to handle `=;` in file names specially

corneliusroemer commented on snakemake/workflow.py:
```diff
@@ -1114,6 +1140,25 @@ def log_provenance_info():
                 if stats:
                     self.scheduler.stats.to_json(stats)
                 logger.logfile_hint()
+                if self.benchmark_all is not None or self.benchmark_output is not None:
```
Benchmarks are currently only written `if success` is true. Would be good to write in any case, I think. While some rules might have failed, it might only be a last rule and benchmark might still be valuable.

SichongP replied:
Updated to gather benchmarks whether success or not

corneliusroemer commented on tests/test_benchmark/expected-results/benchmark.tsv:
```diff
@@ -0,0 +1,6 @@
+jobid	rule	wildcards	resources	input_size_mb	threads	s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load	cpu_time	input_file_size_mb
+2	bench_script	NA	_cores=1;_nodes=1;tmpdir=/tmp	0.0	1	1.1346	0:00:01	32.8	45.8	19.93	23.09	0.0	0.0	0.0	0.09	NA
```
I would maybe move the potentially very long columns to the right hand side:
- resources
- input_size_mb

The most relevant ones could be moved further left:
- CPU
- wallclock time

SichongP replied:
I've moved resources and input_size_mb to right hand side, jobid, rule, and wildcards stay on the left hand side since they serve as "index". Other columnd remain in the same order as original benchmark output

corneliusroemer commented on tests/test_benchmark/expected-results/benchmark.tsv:
```diff
@@ -0,0 +1,6 @@
+jobid	rule	wildcards	resources	input_size_mb	threads	s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load	cpu_time	input_file_size_mb
+2	bench_script	NA	_cores=1;_nodes=1;tmpdir=/tmp	0.0	1	1.1346	0:00:01	32.8	45.8	19.93	23.09	0.0	0.0	0.0	0.09	NA
+4	bench_run_shell	NA	_cores=1;_nodes=1;tmpdir=/tmp	0.0	1	1.0101	0:00:01	156.17	3464.11	119.68	135.93	0.0	0.01	0.0	2.17	NA
+1	bench_shell	NA	_cores=2;_nodes=1;tmpdir=/tmp	0.0	2	10.0008	0:00:10	4.25	19.73	0.38	1.11	0.0	0.0	180.34	19.08	NA
```
Not sure if wildcards should be `NA` if there aren't any or if this should just be empty string

Also, why is `input_size_mb=0.0`, is this how it's supposed to be?

SichongP replied:
NA is easier for a quick visual check in plain text. Empty string is probably best for parsing with another tool. Though Pandas already converts NA to nan automatically. I'm okay either way with this.

SichongP replied:
And yes, those test output files are empty files made with `touch` so they should be 0 in size

SichongP replied:
I updated PR to use empty strings instead. To make plain text easier to read with this change, output has been changed from tsv to csv

corneliusroemer commented on tests/test_benchmark/expected-results/benchmark.tsv:
```diff
@@ -0,0 +1,6 @@
+jobid	rule	wildcards	resources	input_size_mb	threads	s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load	cpu_time	input_file_size_mb
```
I'd maybe also not put `:` in column names and call them: `cpu_time` and `wall_time` or something like this. Units can be looked up in the documentation, no need to state them in the column name.

corneliusroemer replied:
I see, `s` and `h:m:s` are just two ways of formatting the same number. Bit redundant but potentially worth it for ease of viewing. In that case, call it `wall_s` and `wall_hms` or something like that?

SichongP replied:
pretty sure `h:m:s` is what has been used in current benchmark output so keeping this here is mostly for backward compatibility -- if an update is made after an incomplete run, changing this column name would be a breaking change.

corneliusroemer reviewed (commented):
This would be a great addition, I have a few comments for inspiration, feel free to ignore this unsolicited review.

rahedges commented:
Looking forward to this being merged.

rahedges commented:
It would be great if this could be merged.  I am currently using this feature in development but will not enable it in production code until it is in a Snakemake release.

SichongP commented:
Sorry I've been pretty busy lately. I'll start working on refactoring to new structure this weekend

rahedges commented:
> Sorry I've been pretty busy lately. I'll start working on refactoring to new structure this weekend

Thanks!

SichongP commented:
I updated this PR and addressed some of the comments. As discussed above, this PR adds three new CLI arguments:
- `--nobenchmark` disables benchmarking for all rules
- `--benchmark`  writes metrics of all benchmarked rules to a file either given as argument or `.snakemake/benchmarks/workflow_benchmarks_{timestamp}.csv`
- `--benchmark-all`  writes metrics of all rules to a file either given as argument or `.snakemake/benchmarks/workflow_benchmarks_{timestamp}.csv`

fgvieira commented:
I'd avoid adding more command line arguments (that is what I tried to do on #2691), specially when they are not strictly necessary:
1. `--nobenchmark` (disable benchmarking overall): If you do not want to benchmark, then remove the `benchmark` directive from the rule or ignore the benchmark files (the overhead is not that big).
2.  `--benchmark` (collect all benchmarks to a single file): this can be easily achieved with a `find`/`cat`
3. `--benchmark-all` (enable benchmarking to all rules, even those without the directive): if someone wants to benchmark a specific rule, just add the directive (see 1.).

Having all these new benchmarks can lead to a confusing `tsv` file, but we can also make it extension dependent. If `benchmark` file is a `tsv`, then it stays as it is now (backwards compatible); if it is (e.g.) `json`, then we add everything on a json format.

rahedges commented:
If there is low overhead in creating benchmarks why would you need all of the options?   One way to handle it would be to create a global benchmark with a default name and then use the file name if provided.   All of the individual benchmarks would be still be present if one is only interested in a specific task.

SichongP commented:
> 1. `--nobenchmark` (disable benchmarking overall): If you do not want to benchmark, then remove the `benchmark` directive from the rule or ignore the benchmark files (the overhead is not that big).

Surelly it's easier to disable/enable benchmarks at global level than to go and add/remove benchmark directive from dozens of rules. There are several tickets asking specifcially for this feature, as I linked above.

> 2. `--benchmark` (collect all benchmarks to a single file): this can be easily achieved with a `find`/`cat`

Yes, `--benchmark` is perhaps the least useful addition in this PR but IMO it's still helpful to have if it means less biolerplate code. Especially for someone using pre-built workflows without much experience/knowledge in command line.

> 3. `--benchmark-all` (enable benchmarking to all rules, even those without the directive): if someone wants to benchmark a specific rule, just add the directive (see 1.).
 See my response in 1

SichongP commented:
> If there is low overhead in creating benchmarks why would you need all of the options? One way to handle it would be to create a global benchmark with a default name and then use the file name if provided. All of the individual benchmarks would be still be present if one is only interested in a specific task.

Isn't this exactly what this PR does?

rahedges commented:
Is there any idea when this might get merged?   I am using this in development and would like to have a released version of snakemake to integrate it into my release.

raivivek commented:
@johanneskoester Hopefully it's been long enough since last update on this PR that I can re-raise a review request on this. This feature is great and I'm glad it's being developed here. Thanks!

shaman-narayanasamy commented:
Dear all, 

I am using Snakemake v. 8.25.5. I do not see the options discussed in this issue implemented.  Any idea if these features will be available in the near future? It would definitely be helpful to have all steps/rules benchmarked with a single directive.

Cheers,
Shaman

rahedges commented:
> Dear all,
> 
> I am using Snakemake v. 8.25.5. I do not see the options discussed in this issue implemented. Any idea if these features will be available in the near future? It would definitely be helpful to have all steps/rules benchmarked with a single directive.
> 
> Cheers, Shaman

I'm also interested in a timeline.  I've got a benchmarking project starting soon and would like to be able to run a released version of Snakemake .
